src/Stratis.Bitcoin/Base/AsyncWork/AsyncProvider.cs

```diff
@@ -0,0 +1,212 @@
+
+        private ILoggerFactory loggerFactory;
+        private ILogger logger;
+        private ISignals signals;
+        private readonly INodeLifetime nodeLifetime;
+
+        public AsyncProvider(ILoggerFactory loggerFactory, ISignals signals, INodeLifetime nodeLifetime)
+        {
+            this.lockAsyncDelegates = new object();
+            this.lockAsyncLoops = new object();
+
+            this.asyncDelegateWorkers = new Dictionary<IAsyncDelegateDequeuer, AsyncTaskInfo>();
+
+            this.loggerFactory = Guard.NotNull(loggerFactory, nameof(loggerFactory));
+            this.logger = this.loggerFactory.CreateLogger(nameof(AsyncProvider));
+
+            this.signals = Guard.NotNull(signals, nameof(signals));
+            this.nodeLifetime = Guard.NotNull(nodeLifetime, nameof(nodeLifetime));
+        }
+
+        /// <inheritdoc />
+        public IAsyncDelegateDequeuer<T> CreateAndRunAsyncDelegateDequeuer<T>(string friendlyName, Func<T, CancellationToken, Task> @delegate)
+        {
+            AsyncQueue<T> newDelegate;
+
+            lock (this.lockAsyncDelegates)
+            {
+                newDelegate = new AsyncQueue<T>(new AsyncQueue<T>.OnEnqueueAsync(@delegate));
+
+                this.asyncDelegateWorkers.Add(newDelegate, new AsyncTaskInfo(friendlyName));
+            }
+
+            // task will continue with onAsyncDelegateUnhandledException if @delegate had unhandled exceptions
+            newDelegate.ConsumerTask.ContinueWith(this.onAsyncDelegateUnhandledException, newDelegate, TaskContinuationOptions.OnlyOnFaulted);
+
+            // task will continue with onAsyncDelegateCompleted if @delegate completed or was canceled
+            newDelegate.ConsumerTask.ContinueWith(this.onAsyncDelegateCompleted, newDelegate, TaskContinuationOptions.NotOnFaulted);
+
+            return newDelegate;
+        }
+
+        /// <summary>
+        ///  This method is called when delegateTask had an unhandled exception.
+        /// </summary>
+        /// <param name="task">The delegate task.</param>
+        /// <param name="state">The <see cref="IAsyncDelegateDequeuer"/> that's run by the delegateTask</param>
+        private void onAsyncDelegateUnhandledException(Task task, object state)
+        {
+            AsyncTaskInfo delegateInfo;
+            lock (this.lockAsyncDelegates)
+            {
+                if (this.asyncDelegateWorkers.TryGetValue((IAsyncDelegateDequeuer)state, out delegateInfo))
+                {
+                    IAsyncTaskInfoSetter infoSetter;
+                    infoSetter = (IAsyncTaskInfoSetter)delegateInfo;
+                    infoSetter.Exception = task.Exception.GetBaseException();
+                    infoSetter.Status = task.Status;
+                }
+                else
+                {
+                    // Should never happen.
+                    this.logger.LogError("Cannot find the AsyncDelegateInfo related to the faulted task with Id {0}", task.Id);
```

Comment:
Should we not throw in both cases?

Reply:
I don't think so, anyway not here probably.
If we throw here, the task will eat it and will trigger an unhandled exception.

Current implementation handles the unhandled exception and will show (missing part) in console the status of every task currently in faulted state (this was the goal, do you confirm?)

Reply:
Yeah but this is clearly a bug in the component, how do we handle those?

Reply:
we are going to use try/catch inside all users of this component, if we forget it, we'll have evidence on console and when this happens we can then fix it

if we want it to be clearer, then we should stop the node

note that this implementation isn't 100% done, because we have to decide if/what signal and what to do with nodelifetime